torch/_dynamo/guards.py

```diff
@@ -511,6 +521,106 @@ def combine_scopes(left, right):
+        id_to_name_map = {}
+
+        # TensorReferences to size. See PySimInt.TensorReference for more info.
+        size_parts = []
+
+        # TensorReferences to size. See PySimInt.TensorReference for more info.
+        stride_parts = []
+
+        # We should not have a shape env, or guards if we are not in config.dynamic shapes
+        # But check it anyway.
+        if not config.dynamic_shapes:
+            return finished_expressions
+
+        # tensor_check_names is the primary tensor association mechanism in dynamo.
+        # All other guards installations are driven off of it, so these ones will too.
+        for name in tensor_check_names:
+            tensor_id = tensor_check_ids[name]
+            id_to_name_map[tensor_id] = name
+
+            if tensor_id in self.shape_env.id_to_expr:
+                # If we made it here, this tensor_id is relevant to dynamo guard installation
+                # AND was found in the shape_env
+                tensor_ref_list = self.shape_env.id_to_expr[tensor_id]
+                for tensor_ref in tensor_ref_list:
+                    val = str(tensor_ref.expr)
+                    if val in ("0", "1"):
+                        continue
+                    if tensor_ref.kind == "size":
+                        size_parts.append(tensor_ref)
+                    if tensor_ref.kind == "stride":
+                        stride_parts.append(tensor_ref)
+
+        # Extract all the guard elements out of guards
+        # The guard format, atm, uses tuple position 0 for the expression
+        # and tuple position 1 for a negation. Eventually, these will be collapsed together.
+        expression_and_evaluation = [
+            (guard[0], guard[1]) for guard in self.shape_env.guards
+        ]
+        for expression, evaluation in expression_and_evaluation:
+            # Expression to string so we can do string replacement on it.
+            # Before you cry, this whole file generates python code,
+            # So turning an expression into a string is not a code smell, it will
+            # get passed to eval and executed later on down, along with the rest of our guards.
+            # See: py_code in this file.
+            #
+            # Ex: `Eq(s0, s1**2)`
+            expr_as_str = str(expression)
+
+            for tensor_ref in size_parts:
+                # Here, we find any symbols associated with size, and replace them
+                #
+                # Ex: `Eq(s0, s1**2)` becomes `Eq(x.size()[0], s1**2)`
+                expr_as_str = expr_as_str.replace(
+                    str(tensor_ref.expr),
+                    f"{id_to_name_map[tensor_ref.ref_id]}.size()[{tensor_ref.idx}]",
+                )
+
+            for tensor_ref in stride_parts:
+                # Here, we find any symbols associated with stride, and replace them
+                #
+                # Ex: `Eq(x.size()[0], s1**2)` becomes `Eq(x.size()[0], y.stride()[1]**2)`
```

Comment:
When does Eq get turned into `==`. (This is the kind of thing tests would be good for; you want to exercise all of the sympy expressions and make sure they generate valid Python)

Reply:
Eq does not get turned into Eq. It gets stored into the str as `Eq(x.size()[0], y.stride()[1]**2)` - and then we get the `Eq` from sympy. We don't have to do it this way, we can rewrite it to `==` if you prefer.

Reply:
Did this